editor/plugins/sprite_frames_editor_plugin.cpp

```diff
@@ -683,6 +684,36 @@ void SpriteFramesEditor::_animation_add() {
 	animations->grab_focus();
 }
 
+void SpriteFramesEditor::_animation_duplicate() {
+	if ( !frames->has_animation(edited_anim) ) {
+		print_line("Current animation has no frames");
+		return;
+	}
+
+	StringName anim_to_duplicate = edited_anim;
+	int duplicate_frame_count = frames->get_frame_count(anim_to_duplicate);
+	String new_name = String(edited_anim); // Get the animation name to duplicate, cant use the anim_to_duplicate since its a StringName and cannot be modified
+
+	_animation_add();
+
+	for (int i = 0; i < duplicate_frame_count; i++) {
+		frames->add_frame(edited_anim, frames->get_frame(anim_to_duplicate, i));
+	}
+
+	// Rename the duplicated animation
+	int counter = 1;
+	while (frames->has_animation(new_name + " " + itos(counter))) {
+		counter++;
+	}
+	new_name = new_name + "_" + itos(counter);
+	frames->rename_animation(edited_anim,new_name);
+
+	// Update the UI
+	edited_anim = new_name;
+	_update_library();
+	animations->grab_focus();
```

Comment:
It's not undo-redo friendly. Within the `_animation_add` an undo-redo action is created for which `do` creates an animation named like "New Anim[ #id]" which you're renaming in here. But the registered `undo` is to remove "New Anim[ #id]" animation which would fail as that animation was renamed.
I think you could just not call `_animation_add` at all and create the whole duplication logic in here. Something like:

```suggestion
	int counter = 1;
	String name = String(edited_anim) + " " + itos(counter);
	while (frames->has_animation(name)) {
		counter++;
		name = String(edited_anim) + " " + itos(counter);
	}
	
	undo_redo->create_action(TTR("Duplicate Animation"));
	undo_redo->add_do_method(frames, "add_animation", name);	
	for (int i = 0; i < frames->get_frame_count(edited_anim); i++) {
		undo_redo->add_do_method(frames, "add_frame", name, frames->get_frame(edited_anim, i));
	}
	undo_redo->add_undo_method(frames, "remove_animation", name);	
	undo_redo->add_do_method(this, "_update_library");
	undo_redo->add_undo_method(this, "_update_library");

	// Not sure whether this part is needed/wanted (`_add_animation` is doing the same thing).
	List<Node *> nodes;
	_find_anim_sprites(EditorNode::get_singleton()->get_edited_scene(), &nodes, Ref<SpriteFrames>(frames));
	for (Node *E : nodes) {
		String current = E->call("get_animation");
		undo_redo->add_do_method(E, "set_animation", name);
		undo_redo->add_undo_method(E, "set_animation", current);
	}

	edited_anim = name;

	undo_redo->commit_action();
	animations->grab_focus();
```
Also it might be a good idea to detect if the name of the animation being duplicated ends with a number. Currently duplicating "New Anim 1" would create "New Anim 1 1", but "New Anim 2" would probably be more sensible.